### ISSUE TITLE
Convert KyResponse to an interface for better extensibility

### ISSUE THREAD

**Issue by yslpn** [sindresorhus/ky#643]
# Convert KyResponse to an interface for better extensibility

## Current Behavior
Currently, `KyResponse` is defined as a type in the ky library. This makes it difficult to extend or modify the type when users need to add custom properties or methods.

## Proposed Change
Convert `KyResponse` from a type to an interface. This change would allow users to extend the interface as needed without encountering type conflicts.

## Problem Description
When trying to extend the `KyResponse` type, users encounter type errors. For example:

```typescript
declare module "ky" {
  interface HTTPError<T = unknown> {
    response: KyResponse<T> & { data: T };
  }
}

export const apiKy = ky.create({ prefixUrl: PUBLIC_API_URL, credentials: "include" }).extend({
  hooks: {
    beforeError: [
      async (error) => {
        const { response } = error;
        if (response) {
          error.response.data = await response.json();
        }

        return error;
      },
    ],
  },
});

```

This results in the following error:

```
Subsequent property declarations must have the same type. Property 'response' must be of type 'KyResponse<T>', but here has type '{ json: <J = T>() => Promise<J>; } & Response & { data: T; }'.
```

Attempting to redefine `KyResponse` as an interface leads to a different error:

```typescript
declare module "ky" {
  interface KyResponse<T = unknown> extends Response {
    json: <J = T>() => Promise<J>;
    data: T;
  }
}
```

Error:
```
Duplicate identifier 'KyResponse'.ts(2300)
index.d.ts(10, 15): 'KyResponse' was also declared here.
```

## Proposed Solution
Change the definition of `KyResponse` from a type to an interface in the ky library:

```typescript
interface KyResponse<T = unknown> extends Response {
  json: <J = T>() => Promise<J>;
}
```

This change would allow users to extend the interface as needed:

```typescript
declare module "ky" {
  interface KyResponse<T = unknown> {
    data: T;
  }
}
```

## Benefits
1. Improved extensibility for users who need to add custom properties or methods to `KyResponse`.
2. Better TypeScript support for custom extensions to the ky library.
3. Maintains backwards compatibility while providing more flexibility.

## Potential Drawbacks
1. Slight change in how `KyResponse` is defined, which might require updates to documentation.
2. Possible need for minor adjustments in the library's internal implementation.

## Additional Context
This change would be particularly helpful for users who need to add custom error handling or data processing to their ky-based API calls.

**Comment by fitimbytyqi** [comment]
+1

This would be a great benefit, I am converting a huge codebase from custom fetch to ky but I don't wanna wrap every response into ky.get().json<CustomResponse<Type>>()

**Comment by sindresorhus** [comment]
I don't think we should encourage global module augmentation as it can lead to type conflicts and unexpected behavior across your codebase.

Instead, you can create a wrapper type for your specific use case:

```typescript
interface APIError extends HTTPError {
	parsedData?: unknown;
}

const api = ky.extend({
	hooks: {
		beforeError: [
			async error => {
				const apiError = error as APIError;
				apiError.parsedData = await error.response.json();
				return apiError;
			}
		]
	}
});

try {
	await api.get('/endpoint').json();
} catch (error) {
	if (isHTTPError(error)) {
		const data = (error as APIError).parsedData;
	}
}
```

This keeps the types local to where you need them without polluting the global namespace

I will let the other maintainers share their opinion.

**Comment by sholladay** [comment]
I don't love either approach, they can both cause you headaches down the line. But using `as` to force the type to be what you expect is the simpler approach for most situations. And at least it's clear that you are using a custom type. So yeah, I agree with Sindre.

Very open to other ideas, though, if someone comes up with an alternative to consider.
